Update upgrade-guide-from-monitoring-3.x-to-monitoring-4.y.rst (#1773)
scylladb/scylla-monitoring: docs/source/upgrade/upgrade-guide-from-monitoring-3.x-to-monitoring-4.y.rst

Backup
^^^^^^

> A copy of the Prometheus external directory should be made first and used as the data directory for the new monitoring stack. The new monitoring stack uses newer versions of Prometheus and keeping a backup would enable you to rollback to a previous version of Prometheus.

Reviewer: why move to passive voice? I find the original version clearer.
You switch to singular, but we are talking about versions (i.e. 3.x and 4.y)

Author: To avoid the passive voice, we can use the following paragraph:

We suggest to make a copy of Prometheus' external directory and use it as the data directory for the new monitoring stack. The new monitoring stack uses newer versions of Prometheus and keeping a backup would enable you to rollback to a previous version of Prometheus.

> A copy of the Prometheus external directory should be made first and used as the data directory for the new monitoring stack. The new monitoring stack uses newer versions of Prometheus and keeping a backup would enable you to rollback to a previous version of Prometheus.

Reviewer 2: ```suggestion
We suggest making a copy of Prometheus's external directory to use as the data directory for the new version of Monitoring Stack. The new version of Monitoring Stack uses the new version of Prometheus. If you keep a backup of Prometheus's external directory, you can roll back to the previous Prometheus version.
```

To rollback during the testing mode, follow `Killing the new 4.y Monitoring stack in testing mode`_ as explained previously
and the system will continue to operate normally.

> To rollback to version 3.x, run the following commands after you have completed moving to version 4.y (as shown above):

Reviewer: there's two options for rollback depends were you are at in the process, this is when you complete the upgrade,
the new version doesn't say what it should say

Reviewer 2: @lujogre Here you need to restore the previous version because the meaning was changed (and I misunderstood the instructions after the change).

Author: Hello @annastuchlik , thanks for clarification.

Reviewer: This is still wrong, you can roll back at any time, but the way you roll back is different if you have completed moving to 4.0 or if you are still testing.

This is the case after you completed

and you upgraded to Scylla Monitoring 3.8 about three months ago.

If you open the Overview dashboard and look at your entire retention time (in our example 1 year) you will see that while most of the graphs do

> display the data, the latency graphs are missing a period of time.  The latency graph will only show the last three months in our example from the entire year. That nine months gap (12 months minus 3) is what we want to fill with back-filling.

Reviewer: I find this confusing, all those durations are a real world example, it's not always going to be that.

Reviewer 2: @amnonh I would leave it as is, because the entire "Determine the end time" section is an example - and it is emphasized in the first paragraph of that section ("for example, you have a year of retention data") and in the following ones ("in our example 1 year"). For sure, it's clearer than the previous version.
If we want to make it clearer, the content reorganization would help:

Typically, you need to back-fill the recording rules when using a long retention period.

Example:

You have a year of retention data, and you upgraded to ScyllaDB Monitoring 3.8 about three months ago. If you open the Overview dashboard and look at your entire retention time (one year), you will see that while most of the graphs do display the data, the latency graphs are missing a period of time. The latency graph will only show the last three months from the entire year. That nine months gap (12 months minus 3) is what we want to fill with back-filling.

The point in time that the graphs start will be your back-filling end time. Check in the graph for the exact time.

Backfilling Process
-------------------
Backup
^^^^^^
Backup the external directory containing Prometheus data; if something goes wrong, you can revert the changes.

> To complete the process, you must restart Monitoring Stack at least once. You cannot complete the process without providing the path to the external directory with Prometheus data using the ``-d`` command line option. 

Reviewer: you need to use an external data directory, the -d is the way to do it.

Reviewer 2: @amnonh I think this is what the line says. How would you rephrase it?

Reviewer: You cannot complete the process without using Prometheus external directory. You can set Prometheus external directory with the ``-d`` command line option.

Restart the monitoring stack
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> You need to stop the monitoring stack and run the ``start-all.sh`` command with an additional flag (The 365d retention time used here is used only as an example):

Reviewer: The additional flag is "--storage.tsdb.allow-overlapping-blocks" it should be clear

Reviewer 2: Let's add the flag name:

You need to stop the monitoring stack and run the ``start-all.sh`` command with an additional flag `--storage.tsdb.allow-overlapping-blocks`. In the following command, the 365d retention time used is used as an example:

``start-all.sh -d data_dir -b "--storage.tsdb.allow-overlapping-blocks --storage.tsdb.retention.time=365d"``

Create the data files
^^^^^^^^^^^^^^^^^^^^^^^^^

> We will create the data files using the Promtool utility, which has been installed in the Docker container. To run the utility, you must pass the start time and end time in the epoch format. The following example shows one of the ways to convert the times to epoch when the start time is 360 and the end time is 90 days ago:

Reviewer: I would make it clearer, it's insalled in the prometheus docker container

Reviewer 2: "We will create the data files using the Promtool utility, which is installed in the **Prometheus** Docker container". - Is that what you mean?

Author: Hello @annastuchlik / @amnonh  , the paragraph could be:

We will create the data files using the Promtool utility, which is installed in the Prometheus Docker container. To run the utility, you must pass the start time and end time in the epoch format. The following example shows one of the ways to convert the times to epoch when the start time is 360 and the end time is 90 days ago:

Reviewer 2: Sounds good to me.

Author: Hello @annastuchlik , great, thanks for feedback.

> We will create the data files using the Promtool utility, which is installed in the Prometheus Docker container. To run the utility, you must pass the start time and end time in the epoch format. The following example shows one of the ways to convert the times to epoch when the start time is 360 and the end time is 90 days ago:

Reviewer: 360 days ago

Backfilling Process
-------------------
Backup
^^^^^^

> Backup any external directory containing Prometheus data; if something goes wrong, you can revert the changes.

Reviewer: It's not any, it the external data directory

Author: The modified sentence would be:

Backup the external directory containing Prometheus data; if something goes wrong, you can revert the changes.

> Backup any external directory containing Prometheus data; if something goes wrong, you can revert the changes.

Reviewer 2: ```suggestion
Backup the external directory containing Prometheus data - if something goes wrong, you can revert the changes.
```

> The monitoring stack will need to be restarted at least once. This process cannot be completed without an external directory when using the -d command-line option.

Reviewer: This is confusing, see the original description (droping the duplication).
To complete this process, you need to have the data in an exteranl directory, the way you use an external directory is with the -d command line option.

> The monitoring stack will need to be restarted at least once. This process cannot be completed without an external directory when using the -d command-line option.

Reviewer 2: ```suggestion
To complete the process, you must restart Monitoring Stack at least once. You cannot complete the process without providing the path to the external directory with Prometheus data using the ``-d`` command line option. 
```

> To complete the process you will need to restart the monitoring stack at least once. If you are not using an external directory (The ``-d``
> command-line option) You cannot complete it.

Reviewer 2: ```suggestion
```

Create the data files
^^^^^^^^^^^^^^^^^^^^^^^^^

> We will create the data files using the Promtool utility, which has been installed in the docker container. In order to run the utility, the start time and end times must be passed in epoch format. For example, suppose the start and end times are 360 and 90 days back, then the following commands can be used to transform those numbers to epoch: ``echo $((`date +%s` - 3600*24*90))`` and ``echo $((`date +%s` - 3600*24*360))``

Reviewer: This is true only if you are using the docker container.
I prefer the original description that breaks the explenation into stages, you need a start and end time.
you need them in epoch,
there are 100 different ways of transling time to epoch, this is just one example of how to do it from the command line

Author: The new paragraph would be:

We will create the data files using the Promtool utility, which has been installed in the docker container. In order to run the utility, the start time and end times must be passed in epoch format. For example, suppose the start and end times are 360 and 90 days back, then the following commands can be used to transform those numbers to epoch: ``echo $((`date +%s` - 3600*24*90))`` and ``echo $((`date +%s` - 3600*24*360))``  It is important to note that the previous example is just one of many ways to convert time to epoch.

> We will create the data files using the Promtool utility, which has been installed in the docker container. In order to run the utility, the start time and end times must be passed in epoch format. For example, suppose the start and end times are 360 and 90 days back, then the following commands can be used to transform those numbers to epoch: ``echo $((`date +%s` - 3600*24*90))`` and ``echo $((`date +%s` - 3600*24*360))``

Reviewer 2: ```suggestion
We will create the data files using the Promtool utility, which has been installed in the Docker container. To run the utility, you must pass the start time and end time in the epoch format. The following example shows one of the ways to convert the times to epoch when the start time is 360 and the end time is 90 days ago:

``echo $((`date +%s` - 3600*24*360))``

``echo $((`date +%s` - 3600*24*90))``
```

                --url http://localhost:9090 \
                /etc/prometheus/prom_rules/back_fill/3.8/rules.1.yml

> The previous bash script will create a ``data`` directory in the directory where it is executed. The reason to run the bash script under the ``/prometheus/data/`` is to ensure Prometheus has write privileges to the directory.

Reviewer: this is not a bash script

                --url http://localhost:9090 \
                /etc/prometheus/prom_rules/back_fill/3.8/rules.1.yml

> A ``data`` directory will be created in the directory where you run the previous commands. The reason to run it under the ``/prometheus/data/`` is you can be sure Prometheus has write privileges there.

Reviewer: why switch to passive voice? I find the the longer explenation confusing.

Author: Ok

> A ``data`` directory will be created in the directory where you run the previous commands. The reason to run it under the ``/prometheus/data/`` is you can be sure Prometheus has write privileges there.

Reviewer 2: ```suggestion
It will create a ``data`` directory in the directory where you run it. The reason to run it under the ``/prometheus/data/`` is to ensure that Prometheus has write privileges to the directory.
```

    .. note:

>        This process may take a long time depending on the time range and number of cores. For instance, for a cluster with 100 cores, the process took an hour for    every week of data during testing. Hence, please be patient and make sure that the creation process is not interrupted. Please note the time range can be split into smaller intervals (e.g., instead of an entire year, break it down into weeks).

Reviewer: the extra white spaces should be removed

Author: The paragraph without the spaces:

.. note:
       This process may take a long time depending on the time range and number of cores. For instance, for a cluster with 100 cores, the process took an hour for every week of data during testing. Hence, please be patient and make sure that the creation process is not interrupted. Please note the time range can be split into smaller intervals (e.g., instead of an entire year, break it down into weeks).

>        This process may take a long time depending on the time range and number of cores. For instance, for a cluster with 100 cores, the process took an hour for    every week of data during testing. Hence, please be patient and make sure that the creation process is not interrupted. Please note the time range can be split into smaller intervals (e.g., instead of an entire year, break it down into weeks).

Reviewer 2: ```suggestion
       This process may take a long time, depending on the time range and number of cores. For instance, for a cluster with 100 cores, the process took an hour for every week of data during testing. Please be patient and make sure that the creation process is not interrupted. Note that the time range can be split into smaller intervals (e.g., instead of an entire year, break it down into weeks).
```

Copy the data files
^^^^^^^^^^^^^^^^^^^
You should not start this section until all the previous sections have been completed. To copy the data files from the Docker Host to the Prometheus directory in the Docker container, run the following command:

Copy the data files to the Prometheus directory:

.. code-block:: bash

>                 docker cp {options} SRC_PATH aprom:/prometheus/data

Reviewer: remember that in this point in time, you already inside the container at the right directory.
so you're not copying the files from the host, you copy them from the ```data``` directory we created at the previousy step

Author: Understood.
